### DISCUSSION TITLE
access to Marnav sources with nsk_pi-0.2.2/libs/marnav/CMakeLists.txt

### DISCUSSION THREAD

**Discussion by filochard** [nohal/nsk_pi#14] [Ideas]
As you supposed I have some problem to package nsk_pi
Indeed inside : 
opencpn-nsk-plugin-0.2.2/nsk_pi-0.2.2/libs/marnav/CMakeLists.txt
we need to : 

  Download sources and get the source directory
  
  include(FetchContent)
  FetchContent_Declare(
    marnav
    GIT_REPOSITORY ${mnav_repo}
    GIT_TAG ${mnav_tag})
  FetchContent_Populate(marnav)
  FetchContent_GetProperties(marnav SOURCE_DIR marnav_src_dir)

For safety reasons our BuildSystem has no access to download anything from internet when building a rpm : 
we need to provide all the tar.gz sources and the spec file... to our http://svnweb.mageia.org/packages/ which needs a secure access to upload anything

For example I get no problem to have to provide two sources tar.gz : 
dashboardsk_pi-0.2.2.tar.gz
and
opencpn-libs.tar.gz
so that the second fills the extracted Source tree in the right place instead of the opencpn-libs link

![tree2](https://github.com/nohal/nsk_pi/assets/16304019/a2c82f02-4e6c-4ab0-b675-843706ece988)

But no way to download marnav source from github when building the rpm (for security reason !)

Would it be possible to rewrite this CMakeLists.txt and add a link in the source tree
 so that it uses marnav the same way as it uses opencpn-libs

![tree](https://github.com/nohal/nsk_pi/assets/16304019/854816d6-ca25-4241-a163-80f146b49a25)

This would be kind of this : 

![tree3](https://github.com/nohal/nsk_pi/assets/16304019/cc5bfa65-b487-4963-82ad-c1c62eeb1d8c)

I know that it would be a huge work but it will allow a canonical and safe way to build from known and readable opensources better than downloading anything when building

**Comment by nohal** [comment]
Been there already in the early phase of developing this plugin and not likely to do it again. The only alternative I would consider is to look for platform packages, which would in this case mean you would have to package `marnav` as well.

**Comment by nohal** [comment]
https://github.com/nohal/nsk_pi/commit/bab610cf8bb4ff8d214e09feb18efe72c76fca16 is as much as I'm willing to change the logic, if there is `marnav` found on the system, it is used and the plugin build process does not pull in the dependency itself.
It allows you to implement a workflow where you
- upload a marnav tarball
- `cmake .. ; sudo make install` it
- Then build the plugin and it does not need to access the network

**Comment by filochard** [comment]
Hi Pavel
Sorry for this late post (far from home and computer...)
I finally could build a rpm
Only had to modify a little the Plugin.cmake file with this patch (NB the tabs can't be added when copying the patch text)

diff -Nrup a/Plugin.cmake b/Plugin.cmake
--- a/nsk_pi/Plugin.cmakeo	2023-05-20 04:34:15.000000000 +0200
+++ b/Plugin.cmake	2023-05-23 16:26:26.649066399 +0200
@@ -86,8 +86,8 @@
 endmacro()
 
 macro(add_plugin_libraries)
    -  find_path(MARNAV_HEADER marnav)
    -  find_library(MARNAV_LIB libmarnav.a marnav)
    +  find_path(MARNAV_HEADER marnav marnav-io)
    +  find_library(MARNAV_LIB marnav)
                    if("${MARNAV_HEADER}" MATCHES ".*-NOTFOUND" OR MARNAV_LIB MATCHES
                                                                    ".*-NOTFOUND")
     message(STATUS "Pulling in marnav")



to find the installed marnav 


the plugin contains only : 
![rpm-containt](https://github.com/nohal/nsk_pi/assets/16304019/0d69fd3a-d5d9-4434-8dc2-c3efa17d3a36)
Is it right or did I miss anything ?

**Reply by nohal** [reply]
Yes, it is enough given you build against static `marnav` library. To check it you may run `ldd libNSK_pi.so`

**Reply by nohal** [reply]
Or if you build against shared and deploy `libmarnav.so` via a separate package

**Comment by filochard** [comment]
Another question about this plugin : 
Does DashboardSK depends on it or DashboardSK may be installed alone ?

**Reply by nohal** [reply]
DashboardSK does not need it at all, NSK is just a convenience for people who don't have real SK server.

**Comment by filochard** [comment]
Thanks again for your help
I can provide now the only rpm I couldn't build

**Comment by filochard** [comment]
Hi Pavel
I saw your answer in MarioKonrad repo about marnav
I could finally package marnav (without building tools) could install it without conflict

And now since you modified the Plugin.cmake
https://github.com/nohal/nsk_pi/commit/bab610cf8bb4ff8d214e09feb18efe72c76fca16
I could build easily the nsk_pi plugin and package it

Thanks again for your patience and kindness
Philippe

**Comment by filochard** [comment]
Hi again Pavel
I forgot to tell you that I have had to apply a little patch to use a marnav package built upon MarioKonrad source
This patch needs to be applied to Plugin.cmake


diff -Nrup a/Plugin.cmake b/Plugin.cmake
--- a/Plugin.cmake	2023-05-20 04:34:15.000000000 +0200
+++ b/Plugin.cmake	2023-05-23 16:26:26.649066399 +0200
@@ -86,8 +86,8 @@
 endmacro()
 
 macro(add_plugin_libraries)
-  find_path(MARNAV_HEADER marnav)
-  find_library(MARNAV_LIB libmarnav.a marnav)
+  find_path(MARNAV_HEADER marnav marnav-io)
+  find_library(MARNAV_LIB marnav)
   if("${MARNAV_HEADER}" MATCHES ".*-NOTFOUND" OR MARNAV_LIB MATCHES
                                                  ".*-NOTFOUND")
     message(STATUS "Pulling in marnav")


Hope this is useful
Best Regards
Philippe

[nsk_pi-Plugin-cmake.patch.txt](https://github.com/nohal/nsk_pi/files/11663602/nsk_pi-Plugin-cmake.patch.txt)
